### ISSUE TITLE
Distância entre Cep's

### ISSUE THREAD

**Issue by GuilhermeCunha** [filipedeschamps/BrasilAPI#52]
Hoje me deparei com um problema.. e talvez seja do interesse de mais alguém criar uma solução.
Que tal  criar um endpoint que retorne a Distância entre dois Ceps?

**Comment by mukaschultze** [comment]
Seria bem fácil de implementar a distancia se alguém implementasse as coordenadas (#34) primeiro

**Comment by GuilhermeCunha** [comment]
Concordo com você @mukaschultze. Fui dar uma pesquisada de achei a [API do Bing](https://docs.microsoft.com/en-us/bingmaps/rest-services/locations/find-a-location-by-address). 
Que retorna o que a gente precisa através do Endpoint : 
`http://dev.virtualearth.net/REST/v1/Locations?countryRegion=BR&postalCode=$CEP&maxResults=$MAXRESULTS&key=$APIKEY`

Aparentemente a api é Gratuita, mas precisa se registrar para obter a chave da API..

**Comment by VictorGaiva** [comment]
A ideia seria uma distancia cartesiana direta ou uma distancia em rota?

**Comment by JeiselColaco** [comment]
@VictorGaiva, a questão de rota ou distancia, me leva para uma idéia de multiplos CEP's, e o retorno seria na melhor rota para cobrir todos... Sim, problema do caixeiro viajente !!!!

**Comment by VictorGaiva** [comment]
A funcionalidade de distancia entre CEPs por Rota é algo interessante mas extremamente complicadO de implementar, sem usar serviços terceiros como [este oferecido pelo GCP](https://developers.google.com/maps/documentation/distance-matrix/start). Mas ai teria a questão do gerenciamento da API de terceiros, se seria possível fazer cache de resultados, já que dificilmente a mesma rota seria consultada muita vezes, o fato de que rotas podem mudar etc... 

É possível que este projeto entre num Buraco de Coelho se ficar correndo atrás desas funcionalidades sem uma boa direção definida. Acredito que seja interessante dar alguns passos pra trás e pensar sobre quais problemas estamos tentando resolver, Definir o escopo e criar um Roadmap para o projeto.

**Comment by GuilhermeCunha** [comment]
Concordo com o @VictorGaiva, venho tentando achar uma solução para este problema e parece ser bem complicada sem serviços de terceiros mesmo. Agora, poderiam me tirar uma dúvida? 

Tenho estudado o NextJs, para tentar colaborar com projeyo, mas me surgiu uma dúvida.. Conseguimos criar uma api robusta utilizando o NextJs? Em todas as documentações ele se apresenta como um framework **Frontend**..

**Comment by VictorGaiva** [comment]
O sistema seria hospedado no Vercel, que segundo o site também oferece Serverless Functions. Essa solução permite definir funções especificas que rodam sempre que chegar uma requisição em algum endpoint especificado. 

A ideia é hospedar o front-end do projeto no serviço de hospedagem e deixar as funções mais complexas rodando nas Serverless Functions, sendo chamadas pelos endpoints...

**Comment by ronanrodrigo** [comment]
Acho que principalmente a distância em rota não faz sentido pra essa api dado que o domínio desse dado extrapola o contexto da api. O que vocês acham de fazer algo básico apenas para dar a noção de quão perto é um CEP de outro? Isso seria possível com uma lógica aplicada comparando faixa de cep.

Como response, poderia ser um enum de opções que represente o quão próximo em termos de faixa de CEP um está do outro. Por exemplo:

```
{
  faixa: "mesma_sub_regiao
}
```

Com o dado em "faixa" sendo a menor subdivisão em comum ente os dois CEPs. 

Referência
https://www.correios.com.br/enviar/precisa-de-ajuda/tudo-sobre-cep

**Comment by enieber** [comment]
Acredito que esse pedido não faça sentido com base no que o @ronanrodrigo comentou, pois quando vocês busca um cep por distancia isso precisa passar por algumas etapas:

- transformar os CEPs em lat/log
- fazer a busca de distancia entre lat/log

Se um CEP tiver mais de 1 lat/log qual vai usar? como a ideia é que o BrasilAPI não tenha regras de negocio e sim facilite o cache de api, vai ser dificil decidir como fazer essa funcionalidade. O que eu recomendaria é fazer isso na aplicação que for precisar, inclusive poderia fazer a quest dos ceps ao mesmo tempo e diminuir o tempo de esperea.
